src/main/java/org/commonjava/util/partyline/impl/infinispan/model/InfinispanJFS.java

```diff
@@ -0,0 +1,213 @@
+package org.commonjava.util.partyline.impl.infinispan.model;
+
+import com.sun.tools.javac.util.Context;
+import org.commonjava.util.partyline.callback.StreamCallbacks;
+import org.commonjava.util.partyline.lock.local.LocalLockManager;
+import org.commonjava.util.partyline.lock.local.LocalLockOwner;
+import org.commonjava.util.partyline.lock.local.ReentrantOperation;
+import org.commonjava.util.partyline.lock.local.ReentrantOperationLock;
+import org.commonjava.util.partyline.spi.JoinableFile;
+import org.commonjava.util.partyline.spi.JoinableFilesystem;
+import org.infinispan.Cache;
+import org.infinispan.notifications.Listener;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryCreated;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryModified;
+import org.infinispan.notifications.cachelistener.event.CacheEntryEvent;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+
+public class InfinispanJFS
+                implements JoinableFilesystem
+{
+    private final LocalLockManager lockManager = new LocalLockManager();
+
+    private static final long WAIT_TIMEOUT = 100;
+
+    private String nodeKey;
+
+    private final Cache<String, FileMeta> metadataCache;
+
+    private final Cache<UUID, FileBlock> blockCache;
+
+    public InfinispanJFS( final String nodeKey, final Cache<String, FileMeta> metadataCache,
+                          final Cache<UUID, FileBlock> blockCache )
+    {
+        this.nodeKey = nodeKey;
+        this.metadataCache = metadataCache;
+        this.blockCache = blockCache;
+    }
+
+    @Override
+    public JoinableFile getFile( final File file, final LocalLockOwner lockOwner, final StreamCallbacks callbacks,
+                                 final boolean doOutput, final ReentrantOperationLock opLock ) throws IOException
+    {
+        return new InfinispanJF( file, lockOwner, callbacks, doOutput, opLock, this );
+    }
+
+    @Override
+    public LocalLockManager getLocalLockManager()
+    {
+        return null;
+    }
+
+    FileBlock getNextBlock( final FileBlock prevBlock, final FileMeta metadata) throws IOException
+    {
+        synchronized ( InfinispanJFS.this )
+        {
+            // TODO: make sure next will exist
+            UUID next = prevBlock.getNextBlockID();
```

Comment:
I'd put a null check in right below this initialization, then short-circuit and return null if the next UUID is null. That way you don't need to acquire a lock in order to return that null. It would be a faster way to handle EOF.